### ISSUE TITLE
Support for padding around an array?

### ISSUE THREAD

**Issue by ZacCranko** [JuliaArrays/PaddedViews.jl#9]
Is this something that would be useful to support? I had a go at implementing something to this effect this afternoon:

```julia
struct BorderedView{T,N,A} <: AbstractArray{T,N}
    fillvalue::T
    data::A
    loc_1st::NTuple{N,Int}  # first index of data
    loc_end::NTuple{N,Int}  # last index of data
    size::NTuple{N,Int}

    function BorderedView{T,N,A}(fillvalue::T,
                                 data::AbstractArray{T,N},
                                 loc_1st::NTuple{N,Int},
                                 sz::NTuple{N,Int}) where {T,N,A}
        new{T,N,A}(fillvalue, data, loc_1st, loc_1st .+ size(data) .- 1, sz)
    end
end
BorderedView(fillvalue, data::AbstractArray{T,N}, loc_1st, size) where {T,N} = 
    BorderedView{T,N,typeof(data)}(convert(T, fillvalue), data, loc_1st, size)

Base.size(A::BorderedView) = A.size
@inline function Base.getindex(A::BorderedView{T,N}, inds::Vararg{Int,N}) where {T,N}
    @boundscheck 0 .< inds .≤ size(A)

    if any(i < i_1st || i_end < i for (i, i_1st, i_end) in zip(inds, A.loc_1st, A.loc_end))
        return A.fillvalue
    else 
        return A.data[(inds .- A.loc_1st .+ 1)...]
    end
end
```
with:
```
julia> bv = BorderedView(-1, rand(4,4), (2,2), (6,6))
6×6 BorderedView{Float64,2,Array{Float64,2}}:
 -1.0  -1.0       -1.0       -1.0       -1.0        -1.0
 -1.0   0.897366   0.776285   0.483597   0.169369   -1.0
 -1.0   0.171291   0.285309   0.897164   0.222695   -1.0
 -1.0   0.334218   0.437342   0.350231   0.895553   -1.0
 -1.0   0.113691   0.671982   0.772601   0.0831538  -1.0
 -1.0  -1.0       -1.0       -1.0       -1.0        -1.0
```

**Comment by timholy** [comment]
I don't think you need a new type for this, because you can get it by composition:
```julia
julia> using PaddedViews, OffsetArrays

julia> A = OffsetArray(rand(4, 4), 2:5, 2:5)
OffsetArrays.OffsetArray{Float64,2,Array{Float64,2}} with indices 2:5×2:5:
 0.15323    0.219463  0.882128  0.20407 
 0.0992352  0.344492  0.962393  0.73788 
 0.53908    0.585327  0.572761  0.122207
 0.771395   0.443917  0.743549  0.398102

julia> P = PaddedView(-1, A, (Base.OneTo(6), Base.OneTo(6)))
6×6 PaddedViews.PaddedView{Float64,2,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},OffsetArrays.OffsetArray{Float64,2,Array{Float64,2}}}:
 -1.0  -1.0        -1.0       -1.0       -1.0       -1.0
 -1.0   0.15323     0.219463   0.882128   0.20407   -1.0
 -1.0   0.0992352   0.344492   0.962393   0.73788   -1.0
 -1.0   0.53908     0.585327   0.572761   0.122207  -1.0
 -1.0   0.771395    0.443917   0.743549   0.398102  -1.0
 -1.0  -1.0        -1.0       -1.0       -1.0       -1.0
```

Might be worth a mention in the README?

**Comment by ZacCranko** [comment]
It strikes me as a slightly clumsy workaround, but it gets the job done. Definitely worth a mention. I just thought `PaddedViews.jl` would have supported this out of the box.

**Comment by timholy** [comment]
It's not a workaround, it's called orthogonal design :smile:. But one could easily define a convenience function for this. Care to contribute one?

**Comment by ZacCranko** [comment]
I'll have a go at writing one.

**Comment by ZacCranko** [comment]
Is this the kind of thing you had in mind Tim? We will have to add `OffsetArrays` as a dependency. 
Alternately I could tweak the existing `PaddedView` type and `getindex` method to produce the same functionality as above. Or if this is good I'll put together a PR with an updated readme some unit tests.

```julia
using PaddedViews, OffsetArrays

function PaddedView(fillvalue, 
                    data::AbstractArray{T,N}, 
                    data_inds::NTuple{N,AbstractUnitRange}, 
                    array_inds::NTuple{N,AbstractUnitRange}) where {T,N}
    @assert all(map(last, data_inds) .<= map(last, array_inds))
    off_data = OffsetArray(data, data_inds...)
    return PaddedView(fillvalue, off_data, array_inds)
end

function PaddedView(fillvalue, 
                    data::AbstractArray{T,N}, 
                    first_datum::NTuple{N,Integer}, 
                    array_size::NTuple{N,Integer}) where {T,N} 
    data_inds  = map(colon, first_datum, size(data) .+ first_datum .- 1)
    array_inds = map(Base.OneTo, array_size)
    PaddedView(fillvalue, data, data_inds, array_inds)
end
```
```julia
julia> PaddedView(-1, rand(3,3), (2:4,2:4), (1:5,1:5))
PaddedViews.PaddedView{Float64,2,Tuple{UnitRange{Int64},UnitRange{Int64}},OffsetArrays.OffsetArray{Float64,2,Array{Float64,2}}} with indices 1:5×1:5:
 -1.0  -1.0       -1.0       -1.0       -1.0
 -1.0   0.861365   0.391094   0.785754  -1.0
 -1.0   0.186572   0.268956   0.200917  -1.0
 -1.0   0.302953   0.891405   0.205489  -1.0
 -1.0  -1.0       -1.0       -1.0       -1.0

julia> PaddedView(-1, rand(3,3), (2,2), (5,5))
5×5 PaddedViews.PaddedView{Float64,2,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},OffsetArrays.OffsetArray{Float64,2,Array{Float64,2}}}:
 -1.0  -1.0       -1.0        -1.0       -1.0
 -1.0   0.831476   0.69337     0.226009  -1.0
 -1.0   0.366278   0.377273    0.31561   -1.0
 -1.0   0.995109   0.0492788   0.288291  -1.0
 -1.0  -1.0       -1.0        -1.0       -1.0
```

**Comment by timholy** [comment]
I think this is good. Be sure to add a docstring. I will personally probably do it in two stages anyway because that way I don't have to remember which argument is which, but with the docstring it should be pretty clear.

**Comment by timholy** [comment]
Closed by #10
